File: BasicModule/BroadcastStudy/app/src/main/java/com/baiheng/broadcaststudy/MainActivity.java
package com.baiheng.broadcaststudy;

import androidx.appcompat.app.AppCompatActivity;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

/**
 * Broadcast学习的主程序
 */
public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    private static final String TAG = MainActivity.class.getSimpleName();

    private IntentFilter mIntentFilter;
    private IntentFilter mAnotherIntentFilter;
    private IntentFilter mLocalIntentFilter;
    private NetworkChangeReceiver mNetworkChangeReceiver;
    private AnotherBroadcastReceiver mAnotherBroadcastReceiver;
    private LocalReceiver mLocalReceiver;
    private LocalBroadcastManager mLocalBroadcastManager;
    private Button mSendBroadcast;
    private Button mSendOrderBroadcast;
    private Button mSendLocalBroadcast;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mSendBroadcast = (Button) findViewById(R.id.btn_start_broadcast);
        mSendOrderBroadcast = (Button) findViewById(R.id.btn_start_order_broadcast);
        mSendLocalBroadcast = (Button) findViewById(R.id.btn_start_local_broadcast);
        mSendBroadcast.setOnClickListener(this);
        mSendOrderBroadcast.setOnClickListener(this);
        mSendLocalBroadcast.setOnClickListener(this);
        mLocalBroadcastManager = LocalBroadcastManager.getInstance(this);
        registerBroadcast();

    }

    private void registerBroadcast() {
        mIntentFilter = new IntentFilter();
        mIntentFilter.addAction("android.net.conn.CONNECTIVITY_CHANGE");
        mNetworkChangeReceiver = new NetworkChangeReceiver();
        registerReceiver(mNetworkChangeReceiver, mIntentFilter);	//注册广播

        mAnotherIntentFilter = new IntentFilter();
        mAnotherIntentFilter.addAction("com.baiheng.broadcaststudy.MY_BROADCAST");
        mAnotherIntentFilter.setPriority(20);
        mAnotherBroadcastReceiver = new AnotherBroadcastReceiver();
        registerReceiver(mAnotherBroadcastReceiver,mAnotherIntentFilter);

        mLocalIntentFilter = new IntentFilter();
        mLocalIntentFilter.addAction("com.baiheng.broadcaststudy.MY_BROADCAST");
        mLocalReceiver = new LocalReceiver();
        mLocalBroadcastManager.registerReceiver(mLocalReceiver,mLocalIntentFilter);
    }

    @Override
    public void onClick(View view) {
        int id = view.getId();
        switch (id) {
            case R.id.btn_start_broadcast:
                Intent intent = new Intent("com.baiheng.broadcaststudy.MY_BROADCAST");
                intent.setComponent(new ComponentName("com.baiheng.broadcaststudy","com.baiheng.broadcaststudy.MyBroadcastReceiver"));
                sendBroadcast(intent);
                break;
            case R.id.btn_start_order_broadcast:
                Intent intent1 = new Intent("com.baiheng.broadcaststudy.MY_BROADCAST");
                sendOrderedBroadcast(intent1,null);
                break;
            case R.id.btn_start_local_broadcast:
                Intent intent2 = new Intent("com.baiheng.broadcaststudy.MY_BROADCAST");
                mLocalBroadcastManager.sendBroadcast(intent2);
                break;
            default:
                break;
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unregisterReceiver(mNetworkChangeReceiver);
        unregisterReceiver(mAnotherBroadcastReceiver);
        mLocalBroadcastManager.unregisterReceiver(mLocalReceiver);
        Log.d(TAG,"onDestroy");
    }

    class NetworkChangeReceiver extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            ConnectivityManager connectionManager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
            NetworkInfo networkInfo = connectionManager.getActiveNetworkInfo();
            if (networkInfo != null && networkInfo.isConnected()) {
                Toast.makeText(context, "network is Connected", Toast.LENGTH_SHORT).show();
            } else {
                Toast.makeText(context, "network is unavailable", Toast.LENGTH_SHORT).show();
            }
        }
    }

    class LocalReceiver extends BroadcastReceiver{
        @Override
        public void onReceive(Context context, Intent intent) {
            Toast.makeText(context, "received local broadcast", Toast.LENGTH_SHORT).show();
        }
    }
}

File: Multithreading/AIDLStudy/app/src/main/java/com/zhewen/aidlstudy/MainActivity.kt
package com.zhewen.aidlstudy

import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.zhewen.aidlstudy.aidl.CommandClient

class MainActivity : AppCompatActivity(), View.OnClickListener {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        CommandClient.sInstance.init(this)
        findViewById<TextView>(R.id.tv_first).setOnClickListener(this)
        findViewById<TextView>(R.id.tv_second).setOnClickListener(this)
        findViewById<TextView>(R.id.tv_third).setOnClickListener(this)
        findViewById<TextView>(R.id.tv_four).setOnClickListener(this)
        findViewById<TextView>(R.id.tv_five).setOnClickListener(this)
        findViewById<TextView>(R.id.tv_six).setOnClickListener(this)
        CommandClient.sInstance.connectService()
        bindService()
    }


    private val connection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            Log.d("MainActivity","onServiceConnected")
            //绑定服务成功回调
        }

        override fun onServiceDisconnected(name: ComponentName) {
            //服务断开时回调
        }
    }

    private fun bindService() {
        val intent = Intent()
        //Android 5.0开始，启动服务必须使用显示的，不能用隐式的
        intent.action = "com.zhewen.aidlserverstudy.aidl.commandservice"
        intent.`package` = "com.zhewen.aidlserverstudy"
        intent.component = ComponentName("com.zhewen.aidlserverstudy", "com.zhewen.aidlserverstudy.aidl.CommandService")
        val result = bindService(intent, connection, BIND_AUTO_CREATE)
        Log.d("MainActivity", "bindService$result")
    }

    override fun onClick(p0: View?) {
        when(p0?.id) {
            R.id.tv_first -> {
                val result = CommandClient.sInstance.connectService()
                findViewById<TextView>(R.id.tv_first).text = "服务绑定操作结果为:$result"
            }
            R.id.tv_second -> {
               val result =  CommandClient.sInstance.add(10,19)
                findViewById<TextView>(R.id.tv_second).text = "服务端add方法结果——>$result"
            }
            R.id.tv_third -> {
                CommandClient.sInstance.addUserIn()
            }
            R.id.tv_four -> {
                CommandClient.sInstance.addUserOut()
            }
            R.id.tv_five -> {
                CommandClient.sInstance.addUserInOut()
            }
            R.id.tv_six -> {

            }
            else -> {

            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        unbindService(connection)
    }
}